errors.go

```diff
@@ -118,86 +125,138 @@ func (p *Error) LogMetadata() map[string]string {
 // New creates a new error for you. Use this if you want to pass along a custom error code.
 // Otherwise use the handy shorthand factories below
 func New(code string, message string, params map[string]string) *Error {
-	return errorFactory(code, message, params)
+	return errorFactory(nil, code, message, params)
+}
+
+// Propagate takes any error interface and wraps it into an Error.
+// This is useful because an Error contains lots of useful goodies, like the stacktrace of the error.
+// NOTE: If `cause` is already an `Error`, it will merge the message and params with the underlying Error
+func Propagate(cause error, msg string, params map[string]string) error {
+	return PropagateWithCode(cause, ErrInternalService, msg, params)
+}
+
+// PropagateWithCode wraps an error with a custom error code.
+// NOTE: If `cause` is already an `Error`, it will merge the message and params with the underlying Error
+func PropagateWithCode(cause error, code string, msg string, params map[string]string) error {
+	if cause == nil {
+		return nil
+	}
+
+	switch cause := cause.(type) {
+	case *Error:
+		errString := cause.Message
+		if msg != "" {
+			errString = fmt.Sprintf("%s: %s", msg, cause.Message)
+		}
+
+		return &Error{
+			Code:        cause.Code,
```

Comment:
In `WrapWithCode` if you wrapped an existing terror the code you passed in had no effect. I've copied that behaviour in here, but I think that might have been a bug.

It feels like you should be able to _change_ the code by using this method.

Reply:
Honestly the only "Magic" behaviour I would expect from PropagateWithCode is copying the params. Otherwise I'd just expect a nested terror

And I would suggest that Propagete(without code) shouldn't exist

Reply:
I'd also expect the old Wrap functions to propagate a normal error unless there's compelling reason not to change them?

Reply:
The 'code' part kind of confuses me in general. Maybe it's worth laying out some error cases and seeing how we think they should be handled?

1. We have a known failure in our service
Let's say we're trying to debit money but balance is 0. In this case, we should be creating a new terror with some kind of precondition failure code. We propagate that up through the service, adding context but maintaining the code. The framework understands the code and doesn't log, but does serialize with that code.

2. We have an unknown failure in our service.
Let's say it's a recovered panic or some other unexpected error. We receive an error or string from the runtime, and we want to transform it to a terrror with InternalServiceError. To do this, we do `terrors.Wrap` at the call-site, and then we propagate that through the service, log, and serialize.

3. A downstream fails with a non-500 code.
A user is trying to order a card, so we call out to the address service with their userID. For whatever reason, that ID is not found, and the service returns a terror with code NotFound/404. This is not a NotFound for us though, because our action of ordering a card failed, so we want to transform this to an InternalServiceError if we can't handle it (e.g. retries exhausted, or we decide to queue up the order, etc). At this point, we need to distinguish between propagating with changing the code and without.  This cannot be done with the current implementation.

4. A downstream fails with a 500 code.
I think this is actually the same as the above, but the code-change becomes a no-op.

The big departure from existing behaviour here is that in 3 and 4, when you call `terrors.Wrap` on the result of a downstream call, we _don't persist the code_. Error codes die at service boundaries.

So, I think that leads me to believe that the interface should support:
1. Creating a terror with a specific code
1. Propagating an error without changing the code.

If we could break compatibility, I'd suggest that 1. be `terrors.Wrap` and 2 be `terrors.Propagate` and that's it. However, we could introduce a new method like `terrors.Create` to handle 1.

Reply:
These use cases are really helpful 😃 I'm going to use them to see how my proposed changes would hold up.